// @juandardilag/value-objects/dist/StringValueObject/StringValueObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringLengthOptionsDefault = exports.StringValueObject = void 0;
const ValueObject_1 = require("../ValueObject/ValueObject");
const StringValueObjectOperator_1 = require("./StringValueObjectOperator");
const StringValueObjectValidator_1 = require("./StringValueObjectValidator");
class StringValueObject extends ValueObject_1.ValueObject {
    constructor(_value, options) {
        var _a, _b;
        super({
            operable: (options === null || options === void 0 ? void 0 : options.operable) ||
                new StringValueObjectOperator_1.StringValueObjectOperator(((_a = options === null || options === void 0 ? void 0 : options.pii) === null || _a === void 0 ? void 0 : _a.key) || "", ((_b = options === null || options === void 0 ? void 0 : options.pii) === null || _b === void 0 ? void 0 : _b.iv) || ""),
            validatable: (options === null || options === void 0 ? void 0 : options.validatable) ||
                new StringValueObjectValidator_1.StringValueObjectValidator(exports.StringLengthOptionsDefault),
            pii: options === null || options === void 0 ? void 0 : options.pii,
        }, _value);
    }
    static empty() {
        return new StringValueObject("");
    }
    isEmpty() {
        return this._value == "";
    }
    is(value) {
        return this.value === value;
    }
}
exports.StringValueObject = StringValueObject;
exports.StringLengthOptionsDefault = {
    minLength: 0,
    maxLength: 100000,
};
