// @juandardilag/value-objects/dist/NumberValueObject/NumberValueObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NumberValueObject = void 0;
const BooleanValueObject_1 = require("../BooleanValueObject/BooleanValueObject");
const ValueObject_1 = require("../ValueObject/ValueObject");
const NumberValueObjectOperator_1 = require("./NumberValueObjectOperator");
const NumberValueObjectValidator_1 = require("./NumberValueObjectValidator");
class NumberValueObject extends ValueObject_1.ValueObject {
    constructor(_value, _options) {
        var _a, _b;
        super({
            operable: (_a = _options === null || _options === void 0 ? void 0 : _options.operable) !== null && _a !== void 0 ? _a : new NumberValueObjectOperator_1.NumberValueObjectOperator(),
            validatable: (_b = _options === null || _options === void 0 ? void 0 : _options.validatable) !== null && _b !== void 0 ? _b : new NumberValueObjectValidator_1.NumberValueObjectValidator(),
            pii: _options === null || _options === void 0 ? void 0 : _options.pii,
        }, _value);
    }
    empty() {
        return new NumberValueObject(0);
    }
    static zero() {
        return new NumberValueObject(0);
    }
    equalsTo(other) {
        return new BooleanValueObject_1.BooleanValueObject(this._value === other._value);
    }
    differsFrom(other) {
        return new BooleanValueObject_1.BooleanValueObject(this._value !== other._value);
    }
    isLessThan(other) {
        return new BooleanValueObject_1.BooleanValueObject(this._value < other._value);
    }
    isPositive() {
        return new BooleanValueObject_1.BooleanValueObject(this._value > 0);
    }
    isNegative() {
        return new BooleanValueObject_1.BooleanValueObject(this._value < 0);
    }
    isZero() {
        return new BooleanValueObject_1.BooleanValueObject(this._value === 0);
    }
    isBiggerOrEqualThan(other) {
        return new BooleanValueObject_1.BooleanValueObject(this._value > other._value).or(this.equalsTo(other));
    }
    substract(value) {
        return new NumberValueObject(this._value - value._value);
    }
    times(value) {
        return new NumberValueObject(this._value * value._value);
    }
    static from(value) {
        return new NumberValueObject(value._value);
    }
}
exports.NumberValueObject = NumberValueObject;
